### DISCUSSION TITLE
c api 多路径传输

### DISCUSSION THREAD

**Discussion by songyibao** [Tencent/tquic#212] [Q&A (问题和回答)] [answered]
开发者您好，有几个问题想请教您，我是一名新手，如果问题过于基础，还请见谅，感谢！

1. 最新开源版本是否已支持多路径？
（这里的支持多路径是我所理解的支持多网卡并发传输数据吗？）

2. 如果支持多路径
(1) 我查阅到c api文档中有enable_multipath 配置函数，是否启用这个配置就可以透明的实现多路传输？
(2) 我查到文档说多路径有多种调度策略，包括minrtt 或 冗余传输等， 但注意到on_packet_send 回调，文档中说它必须实现， 既然要求应用自行实现，tquic本身的多路径传输和调度是如何使用的呢（多路径相关功能的c api 说明貌似不是很完善，tquic-example-c里也没有）

**Comment by glime-ai** [comment]
您好！很高兴回答您的问题，不用担心问题是否基础，我们都乐于帮助。

1. **最新开源版本是否已支持多路径？**

   是的，根据提供的文档内容，TQUIC 支持多路径传输。这意味着它可以支持多网卡并发传输数据，提高性能及可靠性。

2. **如果支持多路径：**

   - **关于 `enable_multipath` 配置函数：** 是的，启用 `enable_multipath` 配置应该可以让 TQUIC 透明地实现多路传输。这是多路径支持的基础配置，允许一个连接同时使用多个路径。

   - **关于多路径的调度策略和 `on_packet_send` 回调：** TQUIC 支持多种多路径调度策略，如您所提到的 `minrtt` 和冗余传输等。`on_packet_send` 回调的要求意味着在实际发送数据包时，应用程序需要有一定的控制权或者至少需要知道何时数据包被发送。这可能是为了让应用层能够更细粒度地控制数据发送的行为，例如，进行日志记录、监控或实现特定的传输策略。尽管 TQUIC 提供了多路径传输的能力，但具体如何利用这些路径，以及如何调度数据包，可能需要开发者根据自己的应用场景和需求来具体实现或调整。

关于多路径相关功能的 C API 说明和示例不够完善的问题，您可以考虑在 TQUIC 的 GitHub 仓库中提出 issue 或者参与讨论，以获取更详细的指导或建议。同时，持续关注 TQUIC 的官方文档和示例代码更新，可能会有新的示例或文档出现来解答您的疑问。

希望这些回答对您有所帮助！

**Reply by iyangsj** [reply, marked as answer]
`on_packet_send`负责最终将tquic生成的报文发送出去。注意tquic此时已为报文选择了路径，其向`on_packet_send`回调提供的参数除了报文内容本身，还包含元信息，例如发送需使用的源/目的地址等。

**Comment by iyangsj** [comment]
多路径使用示例可以参考：https://github.com/Tencent/tquic/tree/develop/tools
